presentation-api/controlling-ua/reconnectToPresentation_success-manual.html

```diff
@@ -1,72 +1,49 @@
-        return request.start()
-                .then(function (connection) {
-                    // save the presentation id within the
-                    // local storage of the browser
-                    localStorage["presId"] = connection.id;
+    async_test(function (t) {
+        var startBtn = document.getElementById("startBtn");
+        startBtn.onclick = t.step_func(function () {
+            startBtn.disabled = true;
+            var request = new PresentationRequest(presentationUrls);
+            // start presentation
+            request.start().then(t.step_func(function (oldConnection) {
+                // presentationId will be used to reconnect to presentation
+                var presentationId = oldConnection.id;
+                var oldConnectionClosed = false;
+                oldConnection.onclose = t.step_func(function () {
+                    oldConnectionClosed = true;
+                    // reconnect to presentation using presentationId
+                    request.reconnect(presentationId).then(t.step_func(function (newConnection) {
+                        assert_equals(newConnection.state, 'connecting',"connection should be in 'connecting' state");
+                        assert_equals(newConnection.id, presentationId, "Ids of old and new connections must be equal");
+                        t.done();
+                    })).catch(t.step_func(function () {
+                        assert_unreached("Reconnect to presentation promise rejected");
+                    }));
                 });
-    }
-
-    // ------------------------------------------------
-    // Reconnect to Presentation Test (success) - begin
-    // ------------------------------------------------
-    var reconnectToPresentation = function () {
-
-        promise_test(function () {
-
-            // get the saved presentation id
-            // for the reconnect function if exist
-            if (localStorage["presId"]) var presId = localStorage["presId"];
-
-            return request.reconnect(presId)
-                    .then(function (connection) {
-                        assert_true(connection instanceof PresentationConnection);
-                        done();
-                    });
-        }, 'The reconnection was successful.');
-    }
-    // ----------------------------------------------
-    // Reconnect to Presentation Test (success) - end
-    // ----------------------------------------------
+                // close connection. test continue after onclose event is fired.
+                oldConnection.close();
+                // the test fails if after 5s the close event is not fired
+                t.step_timeout(function () {
+                    !oldConnectionClosed && assert_unreached("Close event on old connection not fired");
+                },5000);
+            })).catch(t.step_func(function () {
+                assert_unreached("Error on start presentation");
+            }));
```

Comment:
So, I believe the whole code is correct, but I wonder whether we could improve readability a bit. With the Presentation API, we are in `Promise` land and yet the code needs to catch and assert rejected promises as in the above lines, which would better be caught by the test harness itself. The code also needs to call `step_func` here and there. This makes it hard to read and maintain.

Concretely, I wonder if the following approach could be easier to read and maintain:

```js
function startPresentation() {
        document.getElementById("startBtn").disabled = true;

        promise_test(function (t) {
            var request = new PresentationRequest(presentationUrls);
            var presentationId = null;
            return request.start()
                .then(function (connection) {
                    // No more user input needed, re-enable test timeout
                    t.step_timeout(function () {
                        t.force_timeout();
                        t.done();
                    }, 10000);

                    // Close connection and wait for "close" event
                    return new Promise(function (resolve) {
                        connection.onclose = function () {
                            presentationId = connection.id;
                            resolve();
                        };
                        connection.close();
                    });
                })
                .then(function () {
                    // Connection now closed, let's reconnect to it
                    return request.reconnect(presentationId);
                })
                .then(function (connection) {
                    assert_equals(connection.state, "connecting", "connection should be in 'connecting' state");
                    assert_equals(connection.id, presentationId, "Ids of old and new connections must be equal");
                });
        });
    }
```

I don't think that's perfect either. Using `promise_test` has the advantage of no longer requiring "catch and assert" code. One drawback is the need to create new Promise to wrap the `close` event handler.

(Note I switched to `force_timeout` in that code and use it as a "global" timeout for the test. This is on purpose to avoid looping forever in the "automated" part of the test if e.g. the `reconnect` Promise never gets resolved or rejected. It would be better to have a way to tell the test harness to simply restart the timeout, but I don't think that's possible)

Anyway, this is editorial, feel free to discard. I find these tests hard to write, so I'm trying to find recipes to simplify authoring... but what works for me may not work for others :)